src/topics/lda_model.cpp

```diff
@@ -16,30 +17,23 @@ lda_model::lda_model(const learn::dataset& docs, std::size_t num_topics)
     /* nothing */
 }
 
-void lda_model::save_doc_topic_distributions(const std::string& filename) const
+void lda_model::save_doc_topic_distributions(std::ostream& stream) const
 {
-    std::ofstream file{filename};
-    for (const auto& doc : docs_)
+    io::packed::write(stream, docs_.size());
+    io::packed::write(stream, num_topics_);
+
+    for (const auto& d : docs_)
     {
-        file << doc.id << "\t";
-        double sum = 0;
+        auto dist = topic_distrbution(doc_id{d.id});
         for (topic_id j{0}; j < num_topics_; ++j)
         {
-            double prob = compute_doc_topic_probability(doc.id, j);
-            if (prob > 0)
-                file << j << ":" << prob << "\t";
-            sum += prob;
+            io::packed::write(stream, dist.counts(j));
         }
-        if (std::abs(sum - 1) > 1e-6)
-            throw std::runtime_error{"invalid probability distribution"};
-        file << "\n";
     }
 }
 
-void lda_model::save_topic_term_distributions(const std::string& filename) const
+void lda_model::save_topic_term_distributions(std::ostream& stream) const
```

Comment:
It's the Blei and Lafferty formula, just expanded using log properties to avoid `std::pow`  and potential underflow in the denominator calculation (multiplying lots of small probabilities together might underflow before you can pow and take the log). I would just rename it `bl_term_scorer` and remove the existing `bl_term_scorer` since they should be the same thing (up to the approximation for the log introduced by fastapprox).